### DISCUSSION TITLE
Asyncio: MIssingGreenlet exception on child instance create - no relationships

### DISCUSSION THREAD

**Discussion by karlanke** [sqlalchemy/sqlalchemy#7758] [Usage Questions] [answered]
With the following (simplified) setup:

```py
connection_string = postgresql+asyncpg://....
engine = create_async_engine(connection_string)
session = scoped_session(
        sessionmaker(
            engine, autocommit=False, autoflush=False, expire_on_commit=True, class_=AsyncSession, future=True
        )
    )

class PersistedJob(Base):
    __tablename__ = "jobs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

class PersistedJobPartition(Base):
    __tablename__ = "job_partitions"

    job_id = Column(UUID(as_uuid=True), ForeignKey("jobs.id"), primary_key=True)
    partition_key = Column(String, primary_key=True)
```

I get a MissingGreenlet exception on the following:

```py
stmt = (
      select(PersistedJob)
)

result = await session.execute(stmt)
job = result.scalar()
partition = PersistedJobPartition(
      job_id=job.id,
      partition_key=details.partition_key
)
```

This is a bit simplified - the job fetch is in a different layer, using a transaction wrapper - but I believe it should still be attached to the session. When debugging, the exception triggers on the `job_id=job.id` line specifically. I've tried this a number of ways, with relationships defined with various lazy values, or not at all, and it still seems to happen. Why in the world is SQL getting emitted on *object creation*?

**Comment by CaselIT** [comment, marked as answer]
Hi,

With what you provided it's impossible to guess what's happening.
> When debugging, the exception triggers on the job_id=job.id line specifically. I've tried this a number of ways, with relationships defined with various lazy values, or not at all, and it still seems to happen. Why in the world is SQL getting emitted on object creation?

from this it suggests that `job` is expired, probably because of a commit, so accessing `job.id` makes the orm try to refresh the job from the db, causing the db.

You could try setting `expire_on_commit=False` in the session, or alternatively call `await session.refresh(job)` before using it in that case.

**Reply by zzzeek** [reply]
either one of `job.id` or `details.partition_key` is not loaded from the database, and that's emitting a load.  it should be visible in the stack trace.

**Reply by CaselIT** [reply]
I guess it should get better once https://github.com/sqlalchemy/sqlalchemy/issues/7732 has been implemented

**Reply by karlanke** [reply]
Bah indeed, someone had turned on expire_on_commit, and a separate codepath was doing a commit. Am I correct that the ForeignKey modifier has fewer options to control lazy loading than `relationship`? I was sort of assuming they were orthogonal - ForeignKey controlling schema migrations etc, and `relationship` handling the ORM side of things. I'm not sure how intuitive it is that the mere presence of a FK (required for alembic) causes all this background loading - that may be bias coming from other ORMs, though.

**Reply by zzzeek** [reply]
when the object is expired, it's assumed that the whole row might have been deleted.  so it will load all columns from that state to see that the row wasn't deleted.   this is of course heavy handed for use cases where you're pretty sure nothing like that is going to happen, so assuming that's the case I'd keep expire_on_commit turned off or do explicit refreshing when you move across transaction boundaries.

**Reply by karlanke** [reply]
Quick follow-up with the same example - if I add the following to the partition definition:
```
job: PersistedJob = relationship(
        "PersistedJob",
        backref=backref("partitions", lazy="noload"),
        lazy='noload'
    )
```
everything works great *if* I load at least one job without a join. If, however, I do an immediate:
```
stmt = (
                select(PersistedJob).options(joinedload(PersistedJob.partitions))
                )
```
I get an attribute error on PersistedJob.partitions.

Because I'm running this in an asyncio loop, my impression is that I can never do lazy loads, so I've set up noload everywhere. But, I still want to load relationships in *sometimes*. Is there a better way I should be defining these relationships?

More generally, has there been any thought towards a global flag for "never lazy load"? For performance reasons, I'd love to just set up these relationships to be able to write quick joins, but even if I was in sync-land, I probably wouldn't want surprise DB queries happening in my production setup.

**Reply by karlanke** [reply]
Ah, figured it out - in case anyone else ends up here, I needed to use back_populates, and a relationship on both sides. I am curious about the "never lazy load" idea, though, for production systems.

**Reply by zzzeek** [reply]
well people set up raiseload so that you would never have a lazyload go forward, but the end effect is the same, lazy loading raises an error.   code that is based on async has to make sure eager loading is used up front to load whatever collections will be needed.  there's not a danger to production systems assuming code is fully tested.  if you dont have test suite coverage, that's a bigger problem.
